File: civicBackend/src/models/Complaint.js
// src/models/Complaint.js
// ─────────────────────────────────────────────────────────────
// Complaint Model (Mongoose Schema)
//
// This is the MOST IMPORTANT model. It represents a civic
// complaint submitted by a citizen.
//
// Key design decisions:
//   1. userId: Stores the MongoDB User _id (an ObjectId). This
//      creates a proper "foreign key" relationship. We can use
//      .populate('userId') to fetch user details in one query.
//
//   2. upvotedBy: An ARRAY of User ObjectIds. To check if a
//      user has upvoted, we just check if their ID is in this
//      array. The 'upvotes' field is the count for fast reads.
//      We keep both for performance.
//
//   3. urgency: Initially 'Medium'. After saving, our AI service
//      will update this based on the description text. This is
//      the Phase 2.3 feature.
//
//   4. location: GeoJSON format. This is MongoDB's native
//      geospatial format. Using this allows us to run queries
//      like "find all complaints within 5km of this point".
//      (Phase 2.5 feature)
// ─────────────────────────────────────────────────────────────

const mongoose = require('mongoose');

const complaintSchema = new mongoose.Schema(
  {
    // Reference to the User who filed this complaint
    // ObjectId is MongoDB's auto-generated unique ID type
    userId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User', // Tells Mongoose this links to the 'User' model
      required: [true, 'User ID is required'],
      index: true,
    },
    // Also store the Firebase UID for easy cross-reference
    // without having to join to the User collection every time
    firebaseUid: {
      type: String,
      required: true,
      index: true,
    },
    category: {
      type: String,
      required: [true, 'Category is required'],
      enum: {
        values: ['Sanitation', 'Water', 'Electrical', 'Road', 'Others'],
        message: 'Invalid category. Choose from: Sanitation, Water, Electrical, Road, Others',
      },
    },
    description: {
      type: String,
      required: [true, 'Description is required'],
      trim: true,
      minlength: [10, 'Description must be at least 10 characters long'],
      maxlength: [1000, 'Description cannot exceed 1000 characters'],
    },
    status: {
      type: String,
      enum: {
        values: ['Pending', 'In Progress', 'Resolved', 'Rejected'],
        message: 'Invalid status',
      },
      default: 'Pending',
    },
    urgency: {
      type: String,
      enum: ['Low', 'Medium', 'High'],
      default: 'Medium', // AI will update this after submission
    },
    // Cloudinary image URL
    imageUrl: {
      type: String,
      default: null,
    },
    // Cloudinary voice note URL (Phase 2.3)
    voiceNoteUrl: {
      type: String,
      default: null,
    },
    // Upvote count - denormalized for fast reads (no aggregation needed)
    upvotes: {
      type: Number,
      default: 0,
      min: 0,
    },
    // Array of User ObjectIds who have upvoted this complaint.
    // We use this to prevent double-voting and to undo upvotes.
    upvotedBy: [
      {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'User',
      },
    ],
    // GeoJSON Point format - enables MongoDB's native geo queries
    location: {
      type: {
        type: String,
        enum: ['Point'],
        default: 'Point',
      },
      coordinates: {
        type: [Number], // [longitude, latitude] - NOTE: GeoJSON is lng,lat order
        required: [true, 'Location coordinates are required'],
        validate: {
          validator: function (v) {
            return v.length === 2 && v[0] >= -180 && v[0] <= 180 && v[1] >= -90 && v[1] <= 90;
          },
          message: 'Invalid coordinates. Must be [longitude, latitude]',
        },
      },
    },
    // Status history for auditing
    statusHistory: [
      {
        status: { type: String, required: true },
        changedBy: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true },
        changedAt: { type: Date, default: Date.now },
        note: { type: String, default: '' },
      },
    ],
    // Tracking AI urgency classification status
    urgencyClassification: {
      type: String,
      enum: ['pending', 'done', 'failed'],
      default: 'pending',
    },
  },
  {
    timestamps: true, // Adds createdAt and updatedAt automatically
  }
);

// Create a 2dsphere index on location for geospatial queries.
// This is what makes "find nearby complaints" possible.
complaintSchema.index({ location: '2dsphere' });

module.exports = mongoose.model('Complaint', complaintSchema);

File: civicApp/lib/models/complaint.dart
// lib/models/complaint.dart
// ─────────────────────────────────────────────────────────────
// Complaint Model — maps to the MongoDB Complaint document
// returned by civicBackend.
//
// Key changes from Phase 2.1:
//   - fromMap() is renamed fromJson() for clarity (it parses
//     backend JSON, not Firestore maps)
//   - toJson() is used when POSTING to backend
//   - id uses MongoDB _id field (not Firestore docId)
//   - location is now a GeoJSON object {type, coordinates:[lng,lat]}
//   - Added urgency field (set by AI on backend)
//   - hasUpvoted field is returned per-user by the backend
// ─────────────────────────────────────────────────────────────

class Complaint {
  final String id;
  final String userId;
  final String category;
  final String description;
  final String? imageUrl;
  final String? voiceNoteUrl;
  final String? voiceNoteTranscript;
  final String status;
  final String urgency; // 'Low' | 'Medium' | 'High' — set by AI backend
  final int upvotes;
  final bool hasUpvoted; // Whether current user has upvoted this complaint
  final double lat;
  final double lng;
  final DateTime createdAt;

  const Complaint({
    required this.id,
    required this.userId,
    required this.category,
    required this.description,
    this.imageUrl,
    this.voiceNoteUrl,
    this.voiceNoteTranscript,
    required this.status,
    this.urgency = 'Low',
    required this.upvotes,
    this.hasUpvoted = false,
    required this.lat,
    required this.lng,
    required this.createdAt,
  });

  /// Creates a [Complaint] from the backend's JSON response.
  ///
  /// The backend returns a MongoDB document where:
  ///   - `_id` is the MongoDB ObjectId string
  ///   - `userId` may be populated (an object with name/email) or a raw string
  ///   - `location` is GeoJSON: { type: "Point", coordinates: [lng, lat] }
  factory Complaint.fromJson(Map<String, dynamic> json) {
    // Extract location: GeoJSON coordinates are [longitude, latitude]
    double lat = 0.0;
    double lng = 0.0;
    if (json['location'] != null && json['location']['coordinates'] != null) {
      final coords = json['location']['coordinates'] as List<dynamic>;
      if (coords.length >= 2) {
        lng = (coords[0] as num).toDouble(); // GeoJSON: index 0 = longitude
        lat = (coords[1] as num).toDouble(); // GeoJSON: index 1 = latitude
      }
    }

    // Extract userId: may be a populated object or raw string
    final rawUserId = json['userId'];
    final String userId = rawUserId is Map
        ? (rawUserId['_id'] ?? rawUserId['firebaseUid'] ?? '').toString()
        : rawUserId?.toString() ?? '';

    return Complaint(
      id: json['_id']?.toString() ?? json['id']?.toString() ?? '',
      userId: userId,
      category: json['category']?.toString() ?? 'Others',
      description: json['description']?.toString() ?? '',
      imageUrl: json['imageUrl']?.toString(),
      voiceNoteUrl: json['voiceNoteUrl']?.toString(),
      voiceNoteTranscript: json['voiceNoteTranscript']?.toString(),
      status: json['status']?.toString() ?? 'Pending',
      urgency: json['urgency']?.toString() ?? 'Low',
      upvotes: (json['upvotes'] as num?)?.toInt() ?? 0,
      hasUpvoted: json['hasUpvoted'] as bool? ?? false,
      lat: lat,
      lng: lng,
      createdAt: json['createdAt'] != null
          ? DateTime.tryParse(json['createdAt'].toString()) ?? DateTime.now()
          : DateTime.now(),
    );
  }

  /// Serializes the complaint for a POST /api/complaints request body.
  Map<String, dynamic> toJson() {
    return {
      'category': category,
      'description': description,
      'lat': lat,
      'lng': lng,
      if (imageUrl != null) 'imageUrl': imageUrl,
      if (voiceNoteUrl != null) 'voiceNoteUrl': voiceNoteUrl,
    };
  }

  /// Creates a copy of this complaint with updated fields.
  Complaint copyWith({
    String? status,
    int? upvotes,
    bool? hasUpvoted,
    String? urgency,
  }) {
    return Complaint(
      id: id,
      userId: userId,
      category: category,
      description: description,
      imageUrl: imageUrl,
      voiceNoteUrl: voiceNoteUrl,
      voiceNoteTranscript: voiceNoteTranscript,
      status: status ?? this.status,
      urgency: urgency ?? this.urgency,
      upvotes: upvotes ?? this.upvotes,
      hasUpvoted: hasUpvoted ?? this.hasUpvoted,
      lat: lat,
      lng: lng,
      createdAt: createdAt,
    );
  }
}
